src/controller/user.ts

```diff
@@ -0,0 +1,214 @@
+    @responses({
+        201: { description: 'user created successfully' },
+        400: { description: 'missing parameters, invalid password, validation errors' },
+        401: { description: 'token authorization error' },
+        409: { description: 'user already exists' },
+    })
+    public static async createUser(context: BaseContext): Promise<void> {
+        const userRepository: Repository<User> = getManager().getRepository(User)
+        const missingProperties = requiredProperties(context.request.body, ['name', 'email', 'password', 'dob'])
+
+        if (missingProperties.length > 0)
+            return response(context, 400, `missing_paramters: ${missingProperties.join(', ')}`)
+
+        const passwordValidationResult = validatePassword(context.request.body.password)
+
+        if (passwordValidationResult !== 'good')
+            return response(context, 400, `bad_password: ${passwordValidationResult}`)
+
+        const userToBeSaved: User = new User()
+
+        userToBeSaved.name = context.request.body.name
+        userToBeSaved.email = context.request.body.email
+        userToBeSaved.password = context.request.body.password
+        userToBeSaved.dob = context.request.body.dob
+        userToBeSaved.following = []
+        userToBeSaved.address = context.request.body.address || ''
+        userToBeSaved.description = context.request.body.description || ''
+
+        const validationErrors = await validate(userToBeSaved, userSchema, ['name', 'email', 'dob', 'password'])
+
+        if (validationErrors) return response(context, 400, validationErrors)
+
+        const call = await safeCall(userRepository.findOne({ email: userToBeSaved.email }))
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (call.res) return response(context, 409, 'user_already_exists')
+
+        // convert users dob to date object
+        if (userToBeSaved.dob) userToBeSaved.dob = new Date(userToBeSaved.dob)
+
+        try {
+            await userToBeSaved.hashPassword()
+            await userRepository.save(userToBeSaved)
+        } catch {
+            return response(context, 400, 'error_creating_user')
+        }
+
+        response(context, 201, 'user_created')
+    }
+
+    @request('patch', '/users/{id}')
+    @summary('Update a user')
+    @path({ id: { type: 'string', required: true, description: 'id of the user to update' }, data: { type: 'object', items: {} } })
+    @body(_.omit(userSchema.properties, ['id', 'createdAt', 'updatedAt', 'following']))
+    @responses({
+        200: { description: 'user updated successfully' },
+        400: { description: 'user not found, user already exists, validation errors, user already exists' },
+        401: { description: 'token authorization error' },
+        403: { description: 'not authorized to perform this action' },
+    })
+    public static async updateUser(context: BaseContext): Promise<void> {
```

Comment:
There are a bunch of `400` case in this handler, how do you expect client apps to distinguish these errors?

Reply:
In the controller, along with the status code, the `request.body` returns a status message e.g. `user_not_found`, `user_already_exists` etc that can let the client know what wrong.
e.g.
```ts
if (call.res) return response(context, 400, 'user_already_exists')
```